poly/src/polynomial/multivariate/sparse.rs

```diff
@@ -0,0 +1,408 @@
+
+    /// Outputs an `l`-variate polynomial which is the sum of `l` `d`-degree
+    /// univariate polynomials where each coefficient is sampled uniformly at random.
+    fn rand<R: Rng>(d: usize, l: usize, rng: &mut R) -> Self {
+        let mut random_terms = Vec::new();
+        random_terms.push((F::rand(rng), SparseTerm::new(vec![])));
+        for var in 0..l {
+            for deg in 1..=d {
+                random_terms.push((F::rand(rng), SparseTerm::new(vec![(var, deg)])));
+            }
+        }
+        Self::from_coefficients_vec(l, random_terms)
+    }
+}
+
+impl<'a, 'b, F: Field, T: Term> Add<&'a SparsePolynomial<F, T>> for &'b SparsePolynomial<F, T> {
+    type Output = SparsePolynomial<F, T>;
+
+    fn add(self, other: &'a SparsePolynomial<F, T>) -> SparsePolynomial<F, T> {
+        let mut result = Vec::new();
+        let mut cur_iter = self.terms.iter().peekable();
+        let mut other_iter = other.terms.iter().peekable();
+        // Since both polynomials are sorted, iterate over them in ascending order,
+        // combining any common terms
+        loop {
+            // Peek at iterators to decide which to take from
+            let which = match (cur_iter.peek(), other_iter.peek()) {
+                (Some(cur), Some(other)) => Some((cur.1).cmp(&other.1)),
+                (Some(_), None) => Some(Ordering::Less),
+                (None, Some(_)) => Some(Ordering::Greater),
+                (None, None) => None,
+            };
+            // Push the smallest element to the `result` coefficient vec
+            result.push(match which {
+                Some(Ordering::Less) => cur_iter.next().unwrap().clone(),
+                Some(Ordering::Equal) => {
+                    let other = other_iter.next().unwrap();
+                    let cur = cur_iter.next().unwrap();
+                    (cur.0 + other.0, cur.1.clone())
+                }
+                Some(Ordering::Greater) => other_iter.next().unwrap().clone(),
+                None => break,
+            });
+        }
+        // Remove any zero terms
+        result.retain(|(c, _)| !c.is_zero());
+        SparsePolynomial {
+            num_vars: core::cmp::max(self.num_vars, other.num_vars),
+            terms: result,
+        }
+    }
+}
+
+impl<'a, 'b, F: Field, T: Term> AddAssign<&'a SparsePolynomial<F, T>> for SparsePolynomial<F, T> {
+    fn add_assign(&mut self, other: &'a SparsePolynomial<F, T>) {
+        *self = &*self + other;
+    }
+}
+
+impl<'a, 'b, F: Field, T: Term> AddAssign<(F, &'a SparsePolynomial<F, T>)>
+    for SparsePolynomial<F, T>
+{
+    fn add_assign(&mut self, (f, other): (F, &'a SparsePolynomial<F, T>)) {
+        let other = Self {
+            num_vars: other.num_vars,
+            terms: other
+                .terms
+                .iter()
+                .map(|(coeff, term)| (*coeff * &f, term.clone()))
+                .collect(),
+        };
+        *self = &*self + &other;
+        self.remove_zeros()
+    }
```

Comment:
Doesn't this leave duplicates? (Might not be a problem, I just noticed that you do deduplication elsewhere.)

Reply:
It shouldn't because the implementation of `Add` combines any duplicates